### PR TITLE
Fix MAGN-8827 Revit and Dynamo have crashes after connecting two identical nodes

### PR DESCRIPTION
**Pull request by Randy-Ma**
### Purpose

This is to address MAGN-8827. Originally for the involved graph, on each node, there is a warning message: "Warning: Converting objects to a function pointer is not allowed". The warning message is not easy to understand.

Basically for graphs like this, there is a cyclic dependency which is causing errors. The warning message as mentioned is a kind of run-time warning. But for this case, the compile time warning is easier to understand.

In the warning reporting implementation, currently all run-time warnings take precedence over compile-time warnings. Also for compile time warnings, only warnings which are related to nodes will be reported. That is why the compile time warning is ignored and not reported.

The fix here is to add a warning of cyclic dependency to each node in the cyclic graph. Later cyclic dependency warnings which are produced in compile time, are modified to take precedence over run time warnnings.

image

### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [X] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@aparajit-pratap 

### FYIs

@monikaprabhu

### CODE REVIEW

File: src/DynamoCore/Models/HomeWorkspaceModel.cs [hunk 1]
```diff
@@ -432,11 +432,26 @@ private void OnUpdateGraphCompleted(AsyncTask task)
...
+                string message = string.Empty;
+                if (messages.ContainsKey(warning.Key) &&
+                    warning.Value.Any(w => w.ID == ProtoCore.BuildData.WarningID.kInvalidStaticCyclicDependency))
+                {
+                    messages.Remove(warning.Key);
+                    message = string.Join("\n", warning.Value.
+                        Where(w => w.ID == ProtoCore.BuildData.WarningID.kInvalidStaticCyclicDependency).
+                        Select(w => w.Message));
+                }
+                else
```
Comment (aparajit-pratap) [inline src/DynamoCore/Models/HomeWorkspaceModel.cs]: Should there not be an additional `else if` here like:
```
else if(messages.ContainsKey(warning.Key)) continue;
else
{ ...
```

File: src/Engine/ProtoAssociative/CodeGen.cs [hunk 2]
```diff
@@ -385,9 +385,12 @@ private bool CyclicDependencyTest(ProtoCore.AssociativeGraph.GraphNode node, ref
 
                 string symbol1 = firstNode.updateNodeRefList[0].nodeList[0].symbol.name;
                 string symbol2 = lastNode.updateNodeRefList[0].nodeList[0].symbol.name;
-                string message = String.Format(ProtoCore.Properties.Resources.kInvalidStaticCyclicDependency, symbol1, symbol2);
 
-                core.BuildStatus.LogWarning(ProtoCore.BuildData.WarningID.kInvalidStaticCyclicDependency, message, core.CurrentDSFileName);
+                foreach (var n in dependencyList.GroupBy(x => x.guid).Select(y => y.First()))
```
Comment (aparajit-pratap) [inline src/Engine/ProtoAssociative/CodeGen.cs]: Are you adding this `foreach` loop to show the warning on all nodes involved in the cyclic dependency?

Comment (Randy-Ma) [inline src/Engine/ProtoAssociative/CodeGen.cs]: Exactly. I do not see a proper algorithm to get one node from all the nodes that cause the cyclic dependency

File: src/DynamoCore/Models/HomeWorkspaceModel.cs (hunk 1 above)
Comment (Randy-Ma) [inline src/DynamoCore/Models/HomeWorkspaceModel.cs]: Yes, thanks!

File: src/Engine/ProtoAssociative/CodeGen.cs (hunk 2 above)
Comment (aparajit-pratap) [inline src/Engine/ProtoAssociative/CodeGen.cs]: Can't we simply leave it as it was implemented before?

Comment (Randy-Ma) [inline src/Engine/ProtoAssociative/CodeGen.cs]: No, originally the warning is logged against any nodes. So they will not be reported. We have to choose at least one node.

**Comment by aparajit-pratap** [comment]
@Randy-Ma I think it would be good to have test cases.

File: test/DynamoCoreTests/DynamoDefects.cs [hunk 3]
```diff
@@ -472,5 +472,24 @@ public void EqualEqualTest_ForStrings_Defect6694()
...
+            // check for number of Nodes and Connectors
+            Assert.AreEqual(2, CurrentDynamoModel.CurrentWorkspace.Nodes.Count());
+            Assert.AreEqual(2, CurrentDynamoModel.CurrentWorkspace.Connectors.Count());
+
+            var node1 = CurrentDynamoModel.CurrentWorkspace.NodeFromWorkspace
+                ("385281c8-3177-4359-a348-a3084e16a41a");
+            var node2 = CurrentDynamoModel.CurrentWorkspace.NodeFromWorkspace
+                ("1ad8632e-7ddc-4cc7-bfa5-58cb899a5ddf");
+
+            Assert.IsTrue(node1.ToolTipText.Contains("cyclic"));
```
Comment (aparajit-pratap) [inline test/DynamoCoreTests/DynamoDefects.cs]: You could check for the whole message here: "A cyclic dependency exists between two variables", could you not?

Comment (aparajit-pratap) [inline test/DynamoCoreTests/DynamoDefects.cs]: @Randy-Ma other than the minor comment above, LGTM. Thanks.

**Pull request merged**